### PR TITLE
Tokenization 1.1

### PR DESCRIPTION
**Pull request by lepidopterane-atsmith**
Chat.handleCommand’s error message is more descriptive.

### CODE REVIEW

File: test/codeu/chat/util/TokenizerTest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package codeu.chat.util;
+
+import static org.junit.Assert.*; 
+
+import java.io.IOException;
+import org.junit.Test; 
+   
+public final class TokenizerTest {
```
Comment (aslo) [inline test/codeu/chat/util/TokenizerTest.java:15]: Nice job adding tests! Consider this to be optional feedback: 

It's usually a good idea to test "edge cases" to make sure your Tokenizer behaves as you'd expect in given interesting inputs. Other things you could test, if you feel so inclined: 
* null input
* an empty string as input
* other types of whitespace characters (such as tabs and newlines) as input.

File: test/codeu/chat/TestRunner.java [hunk 2]
```diff
@@ -28,7 +28,8 @@ public static void main(String[] args) {
              codeu.chat.server.RawControllerTest.class,
              codeu.chat.util.TimeTest.class,
              codeu.chat.util.UuidTest.class,
-             codeu.chat.util.store.StoreTest.class
+             codeu.chat.util.store.StoreTest.class,
+             codeu.chat.util.TokenizerTest.class
```
Comment (aslo) [inline test/codeu/chat/TestRunner.java:32]: This file's permissions have been changed to mark it as executable, which is technically not right. I'd recommend changing this file back to its default permissions by running `chmod 0644 test/codeu/chat/TestRunner.java` if you're using Linux or Mac.

File: src/codeu/chat/client/commandline/Panel.java [hunk 3]
```diff
@@ -17,42 +17,394 @@
...
-    void invoke(Scanner line);
-  }
+import codeu.chat.client.core.Context;
+import codeu.chat.client.core.ConversationContext;
+import codeu.chat.client.core.MessageContext;
+import codeu.chat.client.core.UserContext;
+import codeu.chat.util.Tokenizer;
 
-  private final Map<String, Command> commands = new HashMap<>();
+public final class Chat {
```
Comment (aslo) [inline src/codeu/chat/client/commandline/Panel.java:31]: It looks like you copied all of the code `Chat` class into the `Panel.java` file. Was that intentional? I'd recommend changing this file back so that it only contains the updates necessary for the tokenization feature, and only contains the `Panel` public class.  All of the code in the `Chat` class should really stay in `Chat.java`.

File: src/codeu/chat/client/commandline/Chat.java [hunk 4]
```diff
@@ -46,10 +50,18 @@ public Chat(Context context) {
...
-    final String command = tokens.hasNext() ? tokens.next() : "";
+    final List<String> args = new ArrayList<>();
+    final Tokenizer tokenizer = new Tokenizer(line);
+    
+    try {
+      for (String token = tokenizer.next(); token != null; token = tokenizer.next()) {
+        args.add(token);
+      }
+    } catch (IOException e){ 
+      System.out.println("ERROR! IOException caught in Chat.handleCommand().");
```
Comment (aslo) [inline src/codeu/chat/client/commandline/Chat.java:61]: Generally it's not a good idea to reference file names/line numbers/or method names in an error message like this, because in the future, someone might move the code around or rename the method, and forget to update this line. A better way to communicate this information to the user is to use an exception and a stack trace, which will always be up to date with the files and line numbers.

If you want to make sure this information is included with the error message, try replacing this line with something like:

System.out.println("ERROR! IOException caught.");
e.printStackTrace();

File: src/codeu/chat/util/Tokenizer.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+ * @author (Sarah Abowitz) 
+ * @version (V1.0.0 | 5.20.17)
+ */
+
+package codeu.chat.util;
+
+import java.io.IOException;
+import java.lang.StringBuilder;
+
+public final class Tokenizer {
```
Comment (aslo) [inline src/codeu/chat/util/Tokenizer.java:13]: Nice job with the javadoc on this class and its methods! FYI, generally it's not necessary to add javadoc to private methods.

**Review by aslo — CHANGES_REQUESTED** [review]
Looking good! I left a few comments and questions for you that should be addressed before merging this.

Also, this code doesn't compile for me - are you able to compile this locally?

File: src/codeu/chat/client/commandline/Chat.java (hunk 4 above)
Comment (lepidopterane-atsmith) [reply src/codeu/chat/client/commandline/Chat.java:61]: I had something like that originally, but Yuhang wanted me to change it. Thanks for the protip!

File: src/codeu/chat/client/commandline/Panel.java (hunk 3 above)
Comment (lepidopterane-atsmith) [reply src/codeu/chat/client/commandline/Panel.java:31]: I think I understand what was going on here. After I finished my activity, I tried my hand at Version Check, which used Chat and had me changing it in ways I didn't with Tokenization. I wanted to revert those changes as soon as the repo was up. A copy of the original Chat was on my clipboard, so I was able to select and paste over it for the current version of chat, but I must've pasted it here too by accident, sorry.

Comment (lepidopterane-atsmith) [reply src/codeu/chat/client/commandline/Panel.java:31]: This is also probably why it doesn't compile on your end.
